proposals/threads/DataSegmentInitialization.md

```diff
@@ -0,0 +1,135 @@
+# Data segment initialization flag proposal
+
+This page describes a proposal for providing a flag to skip data segment
+initialization when instantiating a module.
+
+## Rationale
+
+Under the current threading proposal, to share a module between multiple
+agents, the module must be instantiated multiple times: once per agent.
+Instantiation initializes linear memory with the contents in the module's data
+segments. If the memory is shared between multiple agents, it will be
+initialized multiple times, potentially overwriting stores that occurred after
+the previous initializations.
+
+For example:
+
+```webassembly
+// The module.
+(module
+  (memory (export "memory") 1)
+
+  ;; Some value used as a counter.
+  (data (i32.const 1) "\0")
+
+  ;; Add one to the counter.
+  (func (export "addOne")
+    (i32.store8
+      (i32.const 0)
+      (i32.add
+        (i32.load8_u (i32.const 0))
+        (i32.const 1)))
+  )
+)
+```
+
+```javascript
+// main.js
+let moduleBytes = ...;
+
+WebAssembly.instantiate(moduleBytes).then(
+  ({instance}) => {
+    // Increment our counter.
+    instance.exports.addOne();
+
+    // Spawn a new Worker.
+    new Worker('worker.js');
+  });
+
+// worker.js
+let moduleBytes = ...;
+
+WebAssembly.instantiate(moduleBytes).then(
+  ({instance}) => {
+    // Oops, our counter has been clobbered.
+  });
+```
+
+This can be worked around by storing the data segments in a separate module
+which is only instantiated once, then exporting this memory to be used by
+another module that contains only code. This works, but it cumbersome since it
+requires two modules where one should be enough.
+
+## Solution 1: Provide instantiation flag to prevent data segment initialization
+
+We provide a third argument in `WebAssembly.instantiate` and the
+`WebAssembly.Instance` constructors. This argument has a type with the
+following IDL:
+
+```webidl
+dictionary WebAssemblyInstantiateOptions {
+  boolean initializeMemory = true;
+}
+```
+
+`WebAssembly.instantiate` is modified as follows:
+
+```
+Promise<WebAssemblyInstantiatedSource>
+  instantiate(BufferSource bytes [, importObject] [, options])
+```
+
+The `WebAssembly.Instance` constructor is modified as follows:
+
+```
+new Instance(moduleObject [, importObject] [, options])
+```
+
+If the `initializeMemory` field's value is `false`, then the memory will not be
+initialized from the contents of the module's data segments.
+
+## Solution 2: Provide an additional initializer expression per data segment
+
+The [binary format for the data section](https://webassembly.github.io/spec/binary/modules.html#data-section)
+currently has a collection of segments, each of which has a memory index, an
+initializer expression for its offset, and its raw data.
+
+Since WebAssembly currently does not allow for multiple memories, the memory
+index must be zero. We can repurpose this field as a flags field and use it to
```

Comment:
Repurpose as in make it impossible to specify multiple memories? Or something else?

Reply:
If it means not allowing multiple memories "forever", that doesn't seem wise. IMHO solution 1 looks to be a better approach.

Reply:
Sorry, I should make this more clear. The intention is that it becomes a flags field, but one of the bits can be used to specify that there exists a non-zero memory index, which would be a separate field. I'll add that to the proposal.